### PR TITLE
rust: update jobserver-rs for rust platform builds

### PR DESCRIPTION
**Pull request by bennofs**
###### Motivation for this change

This fixes the following error that sometimes occurred during rust builds with nix:

```
thread '<unnamed>' panicked at 'failed to acquire jobserver token: Resource temporarily unavailable (os error 11)', src/librustc_trans/back/write.rs:1619:28
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: failed to acquire jobserver token: Resource temporarily unavailable (os error 11)
```

I can cleanup the changes to cargo / rust build support before merging if the general idea for these changes is accepted.

Cargo is now built from the rust source tarball as that's the way upstream recommends, see https://github.com/rust-lang/cargo/commit/5c9665f41c6b4d3b99d3b9f8b48a286f5f154692#commitcomment-27271420 and it also needs the patch so it was easiest to just make it use the same source tree.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/compilers/rust/default.nix [hunk 1]
```diff
@@ -7,33 +7,36 @@
...
-rec {
+  cargoVersion = "0.23.0";
+  src = fetchurl {
+      url = "https://static.rust-lang.org/dist/rustc-${version}-src.tar.gz";
+      sha256 = "1lrzzp0nh7s61wgfs2h6ilaqi6iq89f1pd1yaf65l87bssyl4ylb";
+  };
+  # patches that apply to both cargo and rustc
+  commonPatches = [
+    # this patch can be removed when rustc is upgraded to 1.23
+    ./patches/1.22.1-update-vendored-jobserver.patch
```
Comment (Mic92) [inline pkgs/development/compilers/rust/default.nix]: `rustc` is released since 4. January.

Comment (bennofs) [reply pkgs/development/compilers/rust/default.nix]: Yes but I didn't feel confident enough to do a rustc upgrade. If you want to do that instead, go for it. Also note that this bug affects bootstrapping as well, so you'll probably need to use the 1.23 binaries itself for bootstrapping.

Comment (Mic92) [reply pkgs/development/compilers/rust/default.nix]: Those are necessary now anyway as we have rustc 1.24 now.

Comment (bennofs) [reply pkgs/development/compilers/rust/default.nix]: Okay we can close the PR then

**Comment by Mic92** [comment]
You're changes are included in https://github.com/NixOS/nixpkgs/pull/35030
Looks good so far.

**Pull request closed without merging**